content/src/content/jcr_root/apps/core/wcm/sandbox/commons/ui/clientlibs/panelselect/js/PanelSelector.js

```diff
@@ -0,0 +1,398 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2018 Adobe Systems Incorporated
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+/* global CQ */
+(function($, ns, channel, window, undefined) {
+    "use strict";
+
+    var NS = ".cmp-panelselector";
+
+    var selectors = {
+        indexMarker: ".cmp-panelselector__indexMarker"
+    };
+
+    /**
+     * @typedef {Object} PanelSelectorConfig Represents a Panel Selector configuration object
+     * @property {Granite.author.Editable} editable The [Editable]{@link Granite.author.Editable} against which to create the panel selector
+     * @property {HTMLElement} target The target against which to attach the panel selector UI
+     */
+
+    /**
+     * @class CQ.CoreComponents.PanelSelector
+     * @classdesc A Panel Selector creates a UI consisting of a [Coral.Popover]{@link Coral.Popover} and [Coral.Table]{@link Coral.Table}
+     * which allows selection of panel container items for editing, as well as other operations such as reordering of panels.
+     * @param {PanelSelectorConfig} config The Panel Selector configuration object
+     */
+    CQ.CoreComponents.PanelSelector = ns.util.createClass({
+
+        /**
+         * The Panel Selector configuration Object
+         *
+         * @member {PanelSelectorConfig} CQ.CoreComponents.PanelSelector#_config
+         */
+        _config: {},
+
+        /**
+         * An Object that is used to cache HTMLElement hooks for this Class
```

Comment:
It would be nice to give more information about the hooks (what kind of hooks?), maybe give an example of what is meant